enterprise/server/crypter_service/crypter_service.go

```diff
@@ -6,53 +6,387 @@ import (
+	keyRefreshRetryInterval = 30 * time.Second
+	keyRefreshDeadline      = 25 * time.Second
 )
 
+// Note: there are two types of keys in the cache, one with only groupID set
+// (encryption) and one with all values set (decryption).
+type cacheKey struct {
+	groupID string
+	keyID   string
+	version int
+}
+
+func (ck *cacheKey) String() string {
+	if ck.keyID == "" {
+		return ck.groupID
+	} else {
+		return fmt.Sprintf("%s/%s/%d", ck.groupID, ck.keyID, ck.version)
+	}
+}
+
+type cacheEntry struct {
+	mu                 sync.Mutex
+	keyMetadata        *rfpb.EncryptionMetadata
+	derivedKey         []byte
+	lastUse            time.Time
+	expiresAfter       time.Time
+	lastRefreshAttempt time.Time
+}
+
+type keyCache struct {
+	env   environment.Env
+	dbh   interfaces.DBHandle
+	kms   interfaces.KMS
+	clock clockwork.Clock
+	sf    singleflight.Group
+
+	data sync.Map
+
+	mu               sync.Mutex
+	lastRefreshRun   time.Time
+	activeRefreshOps atomic.Int32
+}
+
+func newKeyCache(env environment.Env, clock clockwork.Clock) (*keyCache, error) {
+	kc := &keyCache{
+		env:   env,
+		dbh:   env.GetDBHandle(),
+		kms:   env.GetKMS(),
+		clock: clock,
+	}
+	return kc, nil
+}
+
+func (c *keyCache) checkCacheEntry(ck cacheKey, ce *cacheEntry) {
+	ce.mu.Lock()
+	defer ce.mu.Unlock()
+
+	// If the expiration is far into the future, don't do anything.
+	if ce.expiresAfter.Sub(c.clock.Now()) > *keyTTL/2 {
+		return
+	}
+
+	// If we reached the expiration and the key was not refreshed, then
+	// remove it from the cache.
+	if c.clock.Now().After(ce.expiresAfter) {
+		c.data.Delete(ck)
+		return
+	}
+
+	// Don't try to extend the life of the key if it hasn't been used recently.
+	if c.clock.Now().Sub(ce.lastUse) > *keyTTL/2 {
+		return
+	}
+
+	// Don't try to refresh the key if we already tried recently.
+	if c.clock.Since(ce.lastRefreshAttempt) < keyRefreshRetryInterval {
+		return
+	}
+
+	c.activeRefreshOps.Inc()
+	go func() {
+		defer c.activeRefreshOps.Dec()
+		ctx, cancel := context.WithTimeout(c.env.GetServerContext(), keyRefreshDeadline)
+		defer cancel()
+		ce.mu.Lock()
+		ce.lastRefreshAttempt = c.clock.Now()
+		ce.mu.Unlock()
+		loadedKey, err := c.refreshKey(ctx, ck)
+		if err == nil {
+			ce.mu.Lock()
+			ce.derivedKey = loadedKey.derivedKey
+			ce.keyMetadata = loadedKey.metadata
+			ce.expiresAfter = c.clock.Now().Add(*keyTTL)
+			ce.mu.Unlock()
+		} else {
+			log.Warningf("could not refresh key %q: %s", ck, err)
+		}
+	}()
+}
+
+func (c *keyCache) startRefresher(quitChan chan struct{}) {
+	go func() {
+		for {
+			select {
+			case <-quitChan:
+				return
+			case <-c.clock.After(keyRefreshScanFrequency):
+				break
+			}
+
+			c.data.Range(func(key, value any) bool {
+				ck := key.(cacheKey)
+				ce := value.(*cacheEntry)
+				c.checkCacheEntry(ck, ce)
+				return true
+			})
+
+			c.mu.Lock()
+			c.lastRefreshRun = c.clock.Now()
+			c.mu.Unlock()
+		}
+	}()
+}
+
+func (c *keyCache) testGetLastRefreshRun() time.Time {
```

Comment:
Did you mean to have these test functions?

Reply:
Ah yup 👍

Reply:
I kind of wrecked my brain trying to think of a less invasive way to test it, but ended up with this slightly unsatisfactory approach.